src/vos/tests/vts_wal.c

```diff
@@ -207,17 +243,676 @@ wal_tst_01(void **state)
+{
+	char *buf = akey->iov_buf;
+
+	if (is_daos_obj_type_set(arg->otype, DAOS_OT_AKEY_UINT64)) {
+		memcpy(buf, &wal_key, sizeof(wal_key));
+		akey->iov_len = akey->iov_buf_len = sizeof(wal_key);
+	} else {
+		akey->iov_len = akey->iov_buf_len =
+			snprintf(buf, arg->akey_size,
+				 "akey=%0*lu", WAL_UPDATE_BUF_NR_SIZE, wal_key);
+	}
+	wal_key++;
+}
+
+static inline void
+wal_dkey_gen(d_iov_t *dkey, struct io_test_args *arg)
+{
+	char *buf = dkey->iov_buf;
+
+	if (is_daos_obj_type_set(arg->otype, DAOS_OT_DKEY_UINT64)) {
+		memcpy(buf, &wal_key, sizeof(wal_key));
+		dkey->iov_len = dkey->iov_buf_len = sizeof(wal_key);
+	} else {
+		dkey->iov_len = dkey->iov_buf_len =
+			snprintf(buf, arg->dkey_size,
+				 "dkey=%0*lu", WAL_UPDATE_BUF_NR_SIZE, wal_key);
+	}
+	wal_key++;
+}
+
+static void
+wal_objs_update_and_fetch(struct io_test_args *arg, daos_epoch_t epoch)
+{
+	daos_epoch_t	ep = epoch;
+	int		obj_nr, dkey_nr, v_nr;
+	int		oidx, didx, aidx, rc;
+	d_iov_t		val_iov;
+	daos_key_t	dkey;
+	daos_recx_t	rex;
+	daos_unit_oid_t	oids[num_keys];
+	char		dkey_buf[UPDATE_DKEY_SIZE];
+	char		akey_buf[UPDATE_AKEY_SIZE];
+	char		update_buf[UPDATE_BUF_SIZE];
+	char		fetch_buf[UPDATE_BUF_SIZE];
+	daos_iod_t	iod;
+	d_sg_list_t	sgl;
+	bool		overwrite;
+
+	wal_key = 1;
+	obj_nr = dkey_nr = v_nr = num_keys;
+
+	memset(&iod, 0, sizeof(iod));
+	memset(&rex, 0, sizeof(rex));
+	memset(&sgl, 0, sizeof(sgl));
+
+	dts_buf_render(update_buf, UPDATE_BUF_SIZE);
+	d_iov_set(&val_iov, update_buf, UPDATE_BUF_SIZE);
+	sgl.sg_nr = 1;
+	sgl.sg_nr_out = 0;
+	sgl.sg_iovs = &val_iov;
+	if (arg->ta_flags & TF_REC_EXT) {
+		iod.iod_type	= DAOS_IOD_ARRAY;
+		iod.iod_size	= UPDATE_REC_SIZE;
+		rex.rx_nr	= UPDATE_BUF_SIZE / UPDATE_REC_SIZE;
+	} else {
+		iod.iod_type	= DAOS_IOD_SINGLE;
+		iod.iod_size	= UPDATE_BUF_SIZE;
+		rex.rx_nr	= 1;
+	}
+
+	iod.iod_recxs		= &rex;
+	iod.iod_nr		= 1;
+	iod.iod_name.iov_buf	= akey_buf;
+	dkey.iov_buf		= dkey_buf;
+
+	overwrite = (arg->ta_flags & TF_OVERWRITE);
+	if (overwrite) {
+		wal_dkey_gen(&dkey, arg);
+		rex.rx_idx = hash_key(&dkey,
+				      is_daos_obj_type_set(arg->otype, DAOS_OT_DKEY_UINT64));
+		wal_akey_gen(&iod.iod_name, arg);
+	}
+
+	/* Update KVs */
+	for (oidx = 0; oidx < obj_nr; oidx++) {
+		arg->oid = oids[oidx] = gen_oid(arg->otype);
+
+		for (didx = 0; didx < dkey_nr; didx++) {
+			if (!overwrite) {
+				wal_dkey_gen(&dkey, arg);
+				rex.rx_idx = hash_key(&dkey,
+					is_daos_obj_type_set(arg->otype, DAOS_OT_DKEY_UINT64));
+			}
+
+			for (aidx = 0; aidx < v_nr; aidx++) {
+				wal_print_buf(update_buf, aidx + v_nr * (didx + dkey_nr * oidx));
+				if (!overwrite)
+					wal_akey_gen(&iod.iod_name, arg);
+
+				rc = io_test_obj_update(arg, ep++, 0, &dkey, &iod, &sgl,
+							NULL, true);
+				assert_rc_equal(rc, 0);
+
+				/* Count */
+				inc_cntr(arg->ta_flags);
+			}
+		}
+		/* Refill VOS file from WAL: reopen pool & container */
+		if (oidx == 0)
+			wal_pool_refill(&arg->ctx);
```

Comment:
Why wal_pool_refill() is called on first object update done? Isn't it supposed to be called after all objects update done?

Reply:
It is not. The WAL replay simulation demonstrated that updating objects would trigger the memory corruption (and likely data corruption at verify) sooner when done just after the replay in case of a faulty replay.